components/script/dom/security/csp.rs

```diff
@@ -279,30 +279,47 @@ impl GlobalCspReporting for GlobalScope {
             return;
         }
         warn!("Reporting CSP violations: {:?}", violations);
+        let source_position_was_provided = source_position.is_some();
         let source_position =
             source_position.unwrap_or_else(compute_scripted_caller_source_position);
         for violation in violations {
-            let (sample, resource) = match violation.resource {
-                ViolationResource::Inline { sample } => (sample, "inline".to_owned()),
-                ViolationResource::Url(url) => (Some(String::new()), url.into()),
+            let (sample, resource, is_inline) = match violation.resource {
+                ViolationResource::Inline { sample } => (sample, "inline".to_owned(), true),
+                ViolationResource::Url(url) => (Some(String::new()), url.into(), false),
                 ViolationResource::TrustedTypePolicy { sample } => {
-                    (Some(sample), "trusted-types-policy".to_owned())
+                    (Some(sample), "trusted-types-policy".to_owned(), true)
                 },
                 ViolationResource::TrustedTypeSink { sample } => {
-                    (Some(sample), "trusted-types-sink".to_owned())
+                    (Some(sample), "trusted-types-sink".to_owned(), true)
                 },
-                ViolationResource::Eval { sample } => (sample, "eval".to_owned()),
-                ViolationResource::WasmEval => (None, "wasm-eval".to_owned()),
+                ViolationResource::Eval { sample } => (sample, "eval".to_owned(), true),
+                ViolationResource::WasmEval => (None, "wasm-eval".to_owned(), true),
             };
+            // Determine source location based on violation type and whether position was provided
+            let (source_file, line_number, column_number) =
+                if source_position_was_provided || is_inline {
```

Comment:
Shouldn't these be both true? E.g. the source position should have only been provided if it was inline. Therefore, I think we should assert that if the resource is URL, no source position is provided. Then only in the URL case we set the filename and for the other ones we use the source position.